### ISSUE TITLE
Consider multidispatch (multimethods) usage

### ISSUE THREAD

**Issue by ivan-kleshnin** [ramda/ramda#1634]
Currently Ramda suffers from 2-nd part of [Expression problem](http://c2.com/cgi/wiki?ExpressionProblem). The code is polluted with ad-hoc type dispatching and is not built to be extensible.

Surely we should rely on interfaces whenever possible. The fact is it's not always possible (especially as we start to support transducers etc.)

Suppose I want to add a support for `Map` and `MapLike` to `assoc` (or rather imitate that as close as possible). 

### Current solution (wrapper)

```js
let R = require("ramda");

// String -> a -> {k: v} -> {k: v}
function assoc(k, v, data) {
  if (data.constructor == Map) {
    let data2 = new Map(data);
    data2.set(k, v);
    return data2;
  } else if (data.constructor == MapLike) {
    // ...
  } 
  // <-- growing dynamic IF...
  else {
    return R.assoc(k, v, data); // fallback to original Ramda's implementation
  }
}

function assocPath(/* ... */) {
  // ...
  // <-- growing dynamic IF...
  // ...
  // fallback to original Ramda's implementation
}

// ...

let m = new Map([["a", "A"], ["b", "B"]]);
console.log(assoc("c", "C", m));
console.log(m);
console.log();

let o = {a: "A", b: "B"};
console.log(assoc("c", "C", o));
console.log(o);
```

**Drawback:** every overriden function contains growing condition (+1 for every new type) with fallback to original Ramda's implementation. 

### Better solution (dispatcher)

Unfortunately, JS does not have function decorators but it's still arguably better:

```js
let R = require("ramda");

// String -> a -> {k: v} -> {k: v}
let assoc = pipe(
  R.dispatchByType([Map], function (k, v, data) {
    let data2 = new Map(data);
    data2.set(k, v);
    return data2;
  }),
  R.dispatchByType([MapLike], function (k, v, data) {
    // ...
  })
  // ...
)(R.assoc);

let assocPath = pipe(/* ... */)(R.assocPath);
```

MultiDispatch-style solutions do not rely on giant IF (conditions are dynamically added) and do not require manual fallback to original Ramda version (it occurs automatically). An additional bonus is that it's open to extension **by design** so conveys expectation to add new conditions (types) better.

**Drawback:** there is no standard (and even common) multidispatch mechanics in JS at this moment.

### Possible implementation of MultiDispatch

https://gist.github.com/ivan-kleshnin/c32b4bd295d7725235b5

**Comment by davidchambers** [comment]
> JS does not have function decorators

JavaScript does not have special *syntax* for decorators, but it certainly has decorators. For example:

```javascript
var requireAuth = function(view) {
  return function(user) {
    if (isAuthenticated(user)) {
      return view.apply(null, Array.prototype.slice.call(arguments, 1));
    } else {
      return loginView();
    }
  };
};

var accountsView = requireAuth(function(user) {
  // Will only get here if user is authenticated.
});
```

**Comment by ivan-kleshnin** [comment]
Yes, I meant special syntax.

**Comment by CrossEye** [comment]
I absolutely agree that the _ad hoc_ nature of the dispatching is not great.

But I'm not so sure that I'd want it to become extensible.  If we did, I would only want to do so by introducing something like [Bilby's enviroments][be], for reasons discussed in https://github.com/ramda/ramda/issues/1037#issuecomment-96050496, _et seq_.  The problem is simply that adding extensibility would break referential transparency.  It doesn't seem a worthwhile trade-off for a library so focused on doing FP correctly.

There is significant disagreement among the core team about the proper degree of dispatching, which is probably why it's so inconsistent now.  Some folks feel that Ramda would be best if it implemented its functions for lists and dispatched everything on other types.  Others think that dispatching should be restricted to just the FantasyLand functions.  There are probably positions in between as well.

I don't know if now is the time to open that debate back up.  Activity among the core folks are down.  I am playing with some techniques, that might make such dispatching significantly more efficient.  But I'm not spending as much time on them as I'd like, and it's not quite clear to me if I could fit that in.

One thing is for certain; all the dispatching, transducing, etc, is making our code much uglier than I'd prefer!

  [be]: http://bilby.brianmckenna.org/#environment

**Comment by ivan-kleshnin** [comment]
@CrossEye, thank you, I see. Yeah, the library gradually incomplicates and the very nature of JS is an unfortunate catalyst. 

> The problem is simply that adding extensibility would break referential transparency. It doesn't seem a worthwhile trade-off for a library so focused on doing FP correctly.

I'm strongly against any state in FP helpers library as well. 
My proposal doesn't imply any of it. Basic decorators don't hold state.

> One thing is for certain; all the dispatching, transducing, etc, is making our code much uglier than I'd prefer!

I think the biggest pain point is inner complexity caused by (premature?) performance optimizations and very high bar for platform compatibility. I'd like to contibute a thing or two but can't find enough time to pass through all that `_inner/_x-something/-omg` layers.

> I don't know if now is the time to open that debate back up.

Maybe not. Let's close this issue for now then. I just wanted to know core team opinion and your answer is pretty comprehensive.

**Comment by CrossEye** [comment]
@ivan-kleshnin:

Sorry, been heads-down on something else.

I misunderstood your proposal, obviously.  I thought you were suggesting that the decorated functions overwrote the Ramda named functions.  I'm more interested if that's not your proposal.  And honestly, this lack of referential transparency is a minor glitch, even if we went the other way.  I just am very happy to be able to say that all Ramda functions are referentially transparent.  That's a big win.  If we had an extensibility technique, I personally would probably use it to overwrite the Ramda functions, or the whole `R` namespace anyway... once at app or library load time.

> I think the biggest pain point is inner complexity caused by (premature?) performance optimizations

Unfortunately many of the performance optimizations are not premature.  They were due to significant performance gaps between what we could achieve and what we thought we should achieve / what other libraries were achieving.

In one very real sense, Ramda is itself just a major performance optimization of an earlier library, [Eweda][ew], which had a very similar API, but which was built using elegant (and poor-performing) recursive implementations rather than our ugly `while` loops.

>  and very high bar for platform compatibility. 

I do think it's time to rethink that.


> I'd like to contibute a thing or two but can't find enough time to pass through all that _inner/_x-something/-omg layers. And I believe I'm not alone in that.

I think the `_internal` bit is here to stay.  But I would like to stop using it as much as we are.  I would rather be eating our own dogfood as much as possible, and there's too much magic going on with what's in `_internal`.  The others I would like to see simplified, and as soon as I finish some time-consuming documentation, I'm going back to see if an approach I'd been trying would actually simplify that. 

   [ew]: https://github.com/CrossEye/eweda
